### PR TITLE
adding idProvider knowledge to filters, security context, session. Adding fkAuthentication filter. Adding session Svc integration. Adding ldap authentication provider.

### PR DESCRIPTION
**Pull request by gauravAshok**
Change logs:
1) There are 3 filters for now.
    **IdpAwareOAuth2ClientAuthenticationProcessingFilter** for oauth2 authentications
    **BasicAuthentiationFilter** for username, password
    **FlipkartWebAuthenticationProcessingFilter** for flipkart login
    
    All filters are registered with a idProvider and is executed when it matches with the idProvider value of request. The request will have the client_id(from which idProvider can be fetched) or idProvider value.
    
2) Added idProvider field in Authentication Objects differentiate between multiple objects. Authentication Objects now implements IdpAware Interface.
    For Flipkart auth object we have **FkWebAuthenticationToken**.

3) Added custom session management.
    **SessionServiceRepository**

### CODE REVIEW

**Comment by gauravAshok** [comment]
@fk-ajain Please review.

File: data-layer/pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
   <parent>
     <groupId>com.flipkart.cp.guardian</groupId>
     <artifactId>fk-cp-guardian</artifactId>
-    <version>1.0-SNAPSHOT</version>
+    <version>1.1-SNAPSHOT</version>
```
Comment (fk-ajain) [inline data-layer/pom.xml]: After repo separation, why is dependency on guardian required? Is that a pending task as of now.

File: data-layer/pom.xml [hunk 2]
```diff
@@ -36,7 +36,7 @@
       <dependency>
         <groupId>com.flipkart.cp.guardian.datamodel</groupId>
         <artifactId>data-model</artifactId>
-        <version>1.0-SNAPSHOT</version>
+        <version>1.1-SNAPSHOT</version>
```
Comment (fk-ajain) [inline data-layer/pom.xml]: Same as above. If there are some classes in guardian which are core to authn, lets move them in a module in this repo

File: data-layer/pom.xml (hunk 1 above)
Comment (gauravAshok) [inline data-layer/pom.xml]: There is no dependency. Its just the previous names. All is in this repo.
Will fix the names in later commit.

File: data-layer/src/main/java/com/flipkart/cp/guardian/datalayer/dataobjects/ClientRegistryDo.java [hunk 3]
```diff
@@ -18,6 +18,7 @@
 	String authorizedGrantTypes;
 	Integer accessTokenValidity;
 	Integer refreshTokenValidity;
+	String identityProviders;
```
Comment (fk-ajain) [inline data-layer/src/main/java/com/flipkart/cp/guardian/datalayer/dataobjects/ClientRegistryDo.java]: How is this plural and yet a string. Shouldn't this be a list? Is this a column name with comma-separated values?

File: data-layer/src/test/resources/create.sql [hunk 4]
```diff
@@ -8,6 +8,7 @@ CREATE TABLE IF NOT EXISTS ClientRegistry
 	authorized_grant_types VARCHAR(256) NOT NULL,
 	access_token_validity INTEGER NOT NULL,
 	refresh_token_validity INTEGER NOT NULL,
+	identity_providers VARCHAR(100),
```
Comment (fk-ajain) [inline data-layer/src/test/resources/create.sql]: lets keep this more, maybe 255.

File: data-layer/src/test/resources/create.sql [hunk 5]
```diff
@@ -55,4 +56,13 @@ CREATE TABLE IF NOT EXISTS ScopeUriRegistry
...
 	PRIMARY KEY(scope, uri)
+);
+
+create table oauth_approvals (
+	userId VARCHAR(256),
+	clientId VARCHAR(256),
+	scope VARCHAR(256),
+	status VARCHAR(10),
+	expiresAt TIMESTAMP,
+	lastModifiedAt TIMESTAMP default NOW()
```
Comment (fk-ajain) [inline data-layer/src/test/resources/create.sql]: composite primary key -> userid, clientid, scope
clientid -> foreign key constraint
scope -> foreign key constraint

File: data-model/src/main/java/com/flipkart/cp/guardian/datamodel/entity/request/CreateClientRequest.java [hunk 6]
```diff
@@ -104,11 +113,24 @@ public void validate() throws RequestValidationException {
...
 			}
 		}
+		
+		/* list of idProvider should not be too long */
+		int idpLength = 0;
+		if(this.getIdentityProviders() != null) {
+			for(String idp : this.getIdentityProviders()) {
+				idpLength += idp.length() + 1;					/* adding 1 for comma separator */
+			}
+			idpLength--;										/* removing last comma */
```
Comment (fk-ajain) [inline data-model/src/main/java/com/flipkart/cp/guardian/datamodel/entity/request/CreateClientRequest.java]: shouldn't you replace this with array.join.length

File: data-model/src/main/java/com/flipkart/cp/guardian/datamodel/entity/request/UpdateClientRequest.java [hunk 7]
```diff
@@ -108,5 +116,18 @@ public void validate() throws RequestValidationException {
...
 			}
 		}
+		
+		/* list of idProvider should not be too long */
+		int idpLength = 0;
+		if(this.getIdentityProviders() != null) {
+			for(String idp : this.getIdentityProviders()) {
+				idpLength += idp.length() + 1;					/* adding 1 for comma separator */
+			}
+			idpLength--;										/* removing last comma */
```
Comment (fk-ajain) [inline data-model/src/main/java/com/flipkart/cp/guardian/datamodel/entity/request/UpdateClientRequest.java]: same as above

File: scripts/create.sql [hunk 8]
```diff
@@ -51,12 +52,21 @@ CREATE TABLE IF NOT EXISTS AuthTokenDetail
...
\ No newline at end of file
+) ENGINE=InnoDB;
+
+create table oauth_approvals (
+	userId VARCHAR(256),
+	clientId VARCHAR(256),
+	scope VARCHAR(256),
+	status VARCHAR(10),
+	expiresAt TIMESTAMP,
+	lastModifiedAt TIMESTAMP default NOW()
```
Comment (fk-ajain) [inline scripts/create.sql]: comments for this added above

File: webservice/pom.xml [hunk 9]
```diff
@@ -46,7 +48,35 @@
...
+				</exclusion>
+				<exclusion>
+					<groupId>org.slf4j</groupId>
+					<artifactId>slf4j-api</artifactId>
+				</exclusion>
+				<exclusion>
+					<artifactId>guava</artifactId>
+					<groupId>com.google.guava</groupId>
+				</exclusion>
+			</exclusions>
```
Comment (fk-ajain) [inline webservice/pom.xml]: See if you can just keep the code required from usercluster in this codebase and do away with this dependency

File: webservice/src/main/java/com/flipkart/cp/guardian/authentication/DefaultIdpAwareObject.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+	}
+	
+	@Override
+	public String getIdProvider() {
+		return idProvider;
+	}
+
+	public T getData() {
+		return data;
+	}
```
Comment (fk-ajain) [inline webservice/src/main/java/com/flipkart/cp/guardian/authentication/DefaultIdpAwareObject.java]: If getData not used anywhere, get rid of that member from this class

File: webservice/src/main/java/com/flipkart/cp/guardian/authentication/entrypoint/BadAuthenticationExceptionEntryPoint.java [hunk 11]
```diff
@@ -26,7 +27,11 @@ public void commence(HttpServletRequest request,
 	    response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
 	    if(authException instanceof BadCredentialsException) {
 	    	response.getOutputStream().println("{\"error\":\"invalid_credentials\",\"error_description\":\"Invalid Credentials given\"}");
-		} else {
+		} 
+	    else if(authException instanceof InsufficientAuthenticationException) {
+	    	response.getOutputStream().println("{\"error\":\"isufficient authentication\",\"error_description\":\"" + authException.getMessage() + "\"}");
```
Comment (fk-ajain) [inline webservice/src/main/java/com/flipkart/cp/guardian/authentication/entrypoint/BadAuthenticationExceptionEntryPoint.java]: typo: insufficient

File: webservice/src/main/java/com/flipkart/cp/guardian/configuration/AuthorizationServerConfiguration.java [hunk 12]
```diff
@@ -55,4 +86,19 @@ public void configure(AuthorizationServerSecurityConfigurer security) throws Exc
 		super.configure(security);
 		security.passwordEncoder(passwordEncoder);
 	}
+
+	@Bean
+	public UserApprovalHandler userApprovalHandler() {
+		ApprovalStoreUserApprovalHandler userApprovalHandler = new ApprovalStoreUserApprovalHandler();
+		userApprovalHandler.setClientDetailsService(clientDetailsService);
+		userApprovalHandler.setApprovalExpiryInSeconds(6 * 30 * 24 * 60 * 60);  // 6 months
```
Comment (fk-ajain) [inline webservice/src/main/java/com/flipkart/cp/guardian/configuration/AuthorizationServerConfiguration.java]: lets pick this from configuration

File: webservice/src/main/java/com/flipkart/cp/guardian/configuration/SwaggerConfiguration.java [hunk 13]
```diff
@@ -12,19 +12,19 @@
 
 /**
  * @author shubham.jain
- *
  */
 @Configuration
 @EnableSwagger2
-public class SwaggerConfig {
+public class SwaggerConfiguration {
```
Comment (fk-ajain) [inline webservice/src/main/java/com/flipkart/cp/guardian/configuration/SwaggerConfiguration.java]: swagger endpoint should not be behind auth
js, css, fonts, images, etc should not behind auth

**Pull request merged**